example/temperature.cpp

```diff
@@ -0,0 +1,143 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <units/origin.h>
+#include <units/physical/si/base/thermodynamic_temperature.h>
+#include <units/physical/si/prefixes.h>
+#include <units/quantity_io.h>
+#include <units/quantity_point.h>
+#include <units/ratio.h>
+#include <iostream>
+
+using namespace units;
+
+namespace si = units::physical::si;
+
+namespace detail {
+
+struct centikelvin : scaled_unit<ratio(1, 100), si::kelvin> {
+};
+
+}  // namespace detail
+
+struct celsius : alias_unit<si::kelvin, basic_symbol_text{"\u00b0C", "`C"}, no_prefix> {
+};
+
+struct celsius_temperature_origin : point_origin<celsius_temperature_origin> {
+  using reference_origin = si::kelvin_temperature_origin;
+  static constexpr auto reference_offset = si::thermodynamic_temperature<::detail::centikelvin, long int>(27315);
+};
+
+template<UnitOf<si::dim_thermodynamic_temperature> U = celsius, QuantityValue Rep = double>
+using celsius_temperature_point = quantity_point<si::dim_thermodynamic_temperature, U, Rep, celsius_temperature_origin>;
+
+struct fahrenheit : named_scaled_unit<fahrenheit, basic_symbol_text{"\u00b0F", "`F"}, no_prefix, ratio(5, 9), si::kelvin> {
+};
+
+namespace detail {
+
+struct millifahrenheit : scaled_unit<ratio(5, 9000), si::kelvin> {
+};
+
+}  // namespace detail
+
+struct fahrenheit_temperature_origin : point_origin<fahrenheit_temperature_origin> {
+  using reference_origin = si::kelvin_temperature_origin;
+  static constexpr auto reference_offset =
+      si::thermodynamic_temperature<::detail::millifahrenheit, long int>(273150 * 9 / 5 + 32000);
+};
+
+template<UnitOf<si::dim_thermodynamic_temperature> U = fahrenheit, QuantityValue Rep = double>
+using fahrenheit_temperature_point =
+    quantity_point<si::dim_thermodynamic_temperature, U, Rep, fahrenheit_temperature_origin>;
+
+
+namespace detail {
+
+template<typename Org>
+concept DerivedPointOrigin = PointOrigin<Org>&& requires
+{
+  typename Org::reference_origin;
```

Comment:
Are there any constraints on the `reference_origin` type itself (i.e. `Quantity`)?

Reply:
Actually, it should be a `PointOrigin` too. Actually, here, the convention is that the `reference_origin` is an origin without any further `reference_origin` set (which makes it some sort of `ReferencePointOrigin`). However, the traits machinery below would probably be simpler if the hypothetical `ReferencePointOrigin` were instead characterised by a recursive `reference_origin` pointing to itself.

In general, this here is a somewhat experimental implementation to enable the implementation of related origins. In this design, an origin is simply a tag identifying the abstract concept of that origin. Then, we need a way to specify the "difference" between related origins, only where it is known and fixed. In order to avoid O(n^2) implementation complexity, currently all related origins need to agree on a reference/base origin.